Avoid closing invalid handles

dylni opened the pull request:
Documentation for [`HandleOrInvalid`] has this note:

> If holds a handle other than `INVALID_HANDLE_VALUE`, it will close the handle on drop.

Documentation for [`HandleOrNull`] has this note:

> If this holds a non-null handle, it will close the handle on drop.

Currently, both will call `CloseHandle` on their invalid handles as a result of using `OwnedHandle` internally, contradicting the above paragraphs. This PR adds destructors that match the documentation.

@rustbot label A-io O-windows T-libs

[`HandleOrInvalid`]: https://doc.rust-lang.org/std/os/windows/io/struct.HandleOrInvalid.html
[`HandleOrNull`]: https://doc.rust-lang.org/std/os/windows/io/struct.HandleOrNull.html

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7234686909/job/19711304054) [(plain)](https://github.com/rust-lang/rust/commit/3cf8d7e23c57c0a491d3631294d1cde0a908251b/checks/19711304054/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_8bb907c4-b010-4489-8bcf-9dae4c42f5bb
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=avoid-closing-invalid-handles
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_8bb907c4-b010-4489-8bcf-9dae4c42f5bb
GITHUB_REF=refs/pull/119029/merge
GITHUB_REF_NAME=119029/merge
GITHUB_REF_PROTECTED=false
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180123 sha256=f323ccf11d14c5b11f6d5e70edb46e391f4b49a5df5c7d922224477ad8ee15c5
  Stored in directory: /tmp/pip-ephem-wheel-cache-iqkzxp94/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
<missing>      4 days ago               /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      4 days ago               /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      4 days ago               /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f115b3d6050>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/9c0bc55f1f9c933417646440b8e46dffc1db26fb5ba79b901066fc8edbab8441a941f974bb2d0529a2965cbd0fff387c9b624c082edc8ee49f3d0dbc5af1ae56
sha256:8291e33b983e90c6b1e66b8177a6f794ceb8b87b29096c973fe30d0355472fc0
---
    Checking object v0.32.1
    Checking std_detect v0.1.5 (/checkout/library/stdarch/crates/std_detect)
    Checking addr2line v0.21.0
error[E0308]: mismatched types
   --> library/std/src/os/windows/io/handle.rs:164:55
161 | / macro_rules! impl_handle_or {
161 | / macro_rules! impl_handle_or {
162 | |     ( $name:ident, $invalid_handle:expr, $error:ident ) => {
163 | |         impl $name {
164 | |             unsafe fn take_owned_handle(&mut self) -> Option<OwnedHandle> {
    | |                       -----------------               ^^^^^^^^^^^^^^^^^^^ expected `Option<OwnedHandle>`, found `()`
    | |                       |
    | |                       implicitly returns `()` as its body has no tail or `return` expression
189 | |     };
190 | | }
    | |_- in this expansion of `impl_handle_or!`
    | |_- in this expansion of `impl_handle_or!`
191 |   impl_handle_or!(HandleOrNull, ptr::null_mut(), NullHandleError);
    |
    |
    = note:   expected enum `core::option::Option<os::windows::io::handle::OwnedHandle>`

error[E0308]: mismatched types
error[E0308]: mismatched types
   --> library/std/src/os/windows/io/handle.rs:164:55
161 | / macro_rules! impl_handle_or {
161 | / macro_rules! impl_handle_or {
162 | |     ( $name:ident, $invalid_handle:expr, $error:ident ) => {
163 | |         impl $name {
164 | |             unsafe fn take_owned_handle(&mut self) -> Option<OwnedHandle> {
    | |                       -----------------               ^^^^^^^^^^^^^^^^^^^ expected `Option<OwnedHandle>`, found `()`
    | |                       |
    | |                       implicitly returns `()` as its body has no tail or `return` expression
189 | |     };
190 | | }
    | |_- in this expansion of `impl_handle_or!`
    | |_- in this expansion of `impl_handle_or!`
191 |   impl_handle_or!(HandleOrNull, ptr::null_mut(), NullHandleError);
192 |   impl_handle_or!(HandleOrInvalid, sys::c::INVALID_HANDLE_VALUE, InvalidHandleError);
    |
    |
    = note:   expected enum `core::option::Option<os::windows::io::handle::OwnedHandle>`

error[E0308]: mismatched types
error[E0308]: mismatched types
   --> library/std/src/os/windows/io/handle.rs:332:14
    |
332 |         Self(OwnedHandle::from_raw_handle(handle))
    |         ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `ManuallyDrop<OwnedHandle>`, found `OwnedHandle`
    |         arguments to this function are incorrect
    |
    |
    = note: expected struct `ManuallyDrop<os::windows::io::handle::OwnedHandle>`
               found struct `os::windows::io::handle::OwnedHandle`
note: tuple struct defined here
   --> library/std/src/os/windows/io/handle.rs:95:12
    |
95  | pub struct HandleOrNull(ManuallyDrop<OwnedHandle>);

error[E0308]: mismatched types
error[E0308]: mismatched types
   --> library/std/src/os/windows/io/handle.rs:355:14
    |
355 |         Self(OwnedHandle::from_raw_handle(handle))
    |         ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `ManuallyDrop<OwnedHandle>`, found `OwnedHandle`
    |         arguments to this function are incorrect
    |
    |
    = note: expected struct `ManuallyDrop<os::windows::io::handle::OwnedHandle>`
               found struct `os::windows::io::handle::OwnedHandle`
note: tuple struct defined here
   --> library/std/src/os/windows/io/handle.rs:114:12
    |
114 | pub struct HandleOrInvalid(ManuallyDrop<OwnedHandle>);

For more information about this error, try `rustc --explain E0308`.
error: could not compile `std` (lib) due to 4 previous errors
Build completed unsuccessfully in 0:00:14
```

</details>

thomcc commented:
I'm going to be away for a few months, so I'm rerolling my PRs so that folks don't have to wait for me. Sorry/thanks.

r? libs

joboet commented on library/std/src/os/windows/io/handle.rs:
```diff
@@ -91,7 +92,7 @@ pub struct OwnedHandle {
 #[stable(feature = "io_safety", since = "1.63.0")]
 #[derive(Debug)]
-pub struct HandleOrNull(OwnedHandle);
+pub struct HandleOrNull(ManuallyDrop<OwnedHandle>);
```
I'd simply use `RawHandle` here, so that we don't have to do the `ManuallyDrop::take` dance below.

dylni replied:
Done. Since it's not possible to move out the inner field with a custom `Drop` implementation, I was looking for a decent way to implement `TryFrom`. However, I can just forget the struct altogether.

joboet commented on library/std/src/os/windows/io/handle.rs:
```diff
@@ -157,24 +158,40 @@ impl BorrowedHandle<'_> {
-#[stable(feature = "io_safety", since = "1.63.0")]
-impl TryFrom<HandleOrNull> for OwnedHandle {
-    type Error = NullHandleError;
+macro_rules! impl_handle_or {
```
Especially when combined with the suggestion above, I don't think that a macro is really necessary here.

dylni replied:
I would likely still use a macro due to the shared code between the implementations, but I don't have strong opinions if the preference is to avoid macros here.

joboet reviewed (approved):
That's a very good catch! Though I think this could be written a bit simpler without `ManuallyDrop`.

ChrisDenton commented:
Hi @dylni could you address the comments? Use `@rustbot ready` when you're done, thanks!

@rustbot author

dylni commented:
@rustbot ready

joboet reviewed (approved):
Looks great!

ChrisDenton commented:
Thanks!

@bors r+ rollup

bors commented:
:pushpin: Commit 24556e02826ae3d58ad1f88c186215e72a86a1df has been approved by `ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ChrisDenton 24556e02826ae3d58ad1f88c186215e72a86a1df -->
<!-- homu: {"type":"Approved","sha":"24556e02826ae3d58ad1f88c186215e72a86a1df","approver":"ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

klensy commented on library/std/src/os/windows/io/handle.rs:
```diff
@@ -163,15 +163,24 @@ impl TryFrom<HandleOrNull> for OwnedHandle {
+    }
+}
+
+#[stable(feature = "handle_or_drop", since = "1.75.0")]
```
From where that 1.75?

ChrisDenton replied:
Hm, you're right. And this doesn't need a new feature as the fact it needs dropping isn't itself new.

ChrisDenton commented:
@bors r-

dylni commented:
Updated and rebased on the latest master

```diff
diff --git a/library/std/src/os/windows/io/handle.rs b/library/std/src/os/windows/io/handle.rs
index 5566f525ea4..30a809daa21 100644
--- a/library/std/src/os/windows/io/handle.rs
+++ b/library/std/src/os/windows/io/handle.rs
@@ -173,7 +173,7 @@ fn try_from(handle_or_null: HandleOrNull) -> Result<Self, NullHandleError> {
     }
 }
 
-#[stable(feature = "handle_or_drop", since = "1.75.0")]
+#[stable(feature = "io_safety", since = "1.78.0")]
 impl Drop for HandleOrNull {
     #[inline]
     fn drop(&mut self) {
@@ -251,7 +251,7 @@ fn try_from(handle_or_invalid: HandleOrInvalid) -> Result<Self, InvalidHandleErr
     }
 }
 
-#[stable(feature = "handle_or_drop", since = "1.75.0")]
+#[stable(feature = "io_safety", since = "1.78.0")]
 impl Drop for HandleOrInvalid {
     #[inline]
     fn drop(&mut self) {
```

@rustbot ready

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8215859584/job/22469777712) [(plain)](https://github.com/rust-lang/rust/commit/f3b5b086323b3d7c965368753fc1887b9825a0b6/checks/22469777712/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_12a37463-b059-43fa-9ce5-14a004f3c93d
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=avoid-closing-invalid-handles
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_12a37463-b059-43fa-9ce5-14a004f3c93d
GITHUB_REF=refs/pull/119029/merge
GITHUB_REF_NAME=119029/merge
GITHUB_REF_PROTECTED=false
---
#13 3.605 Building wheels for collected packages: reuse
#13 3.605   Building wheel for reuse (pyproject.toml): started
#13 3.937   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.938   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.944   Stored in directory: /tmp/pip-ephem-wheel-cache-w39mox65/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.944 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.965   Attempting uninstall: setuptools
#13 3.966     Found existing installation: setuptools 59.6.0
#13 3.967     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
    Checking object v0.32.2
    Checking std_detect v0.1.5 (/checkout/library/stdarch/crates/std_detect)
    Checking miniz_oxide v0.7.2
    Checking addr2line v0.21.0
error[E0711]: feature `io_safety` is declared stable since 1.78.0, but was previously declared stable since 1.63.0
   --> library/std/src/os/windows/io/handle.rs:176:1
    |
176 | #[stable(feature = "io_safety", since = "1.78.0")]


error[E0711]: feature `io_safety` is declared stable since 1.78.0, but was previously declared stable since 1.63.0
   --> library/std/src/os/windows/io/handle.rs:254:1
    |
254 | #[stable(feature = "io_safety", since = "1.78.0")]

For more information about this error, try `rustc --explain E0711`.
error: could not compile `std` (lib) due to 2 previous errors
Build completed unsuccessfully in 0:00:14
```

</details>

dylni committed:
Avoid closing invalid handles

ChrisDenton commented:
Thanks!

@bors r+ rollup

bors commented:
:pushpin: Commit a82587c1d44ac94fb95fb40989bcc2536fd13b9f has been approved by `ChrisDenton`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=ChrisDenton a82587c1d44ac94fb95fb40989bcc2536fd13b9f -->
<!-- homu: {"type":"Approved","sha":"a82587c1d44ac94fb95fb40989bcc2536fd13b9f","approver":"ChrisDenton","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
